### PR TITLE
core, eth: dump dirty snapshot for each transaction

### PR DESCRIPTION
**Pull request by tailingchen**

### CODE REVIEW

File: core/state/dump.go [hunk 1]
```diff
@@ -81,33 +81,81 @@ func (self *StateDB) Dump() []byte {
...
+func newDirtyDump(trie Trie) *DirtyDump {
+	return &DirtyDump{
+		Root:     fmt.Sprintf("%x", trie.Hash()),
+		Accounts: make(map[string][]DirtyDumpChange),
+	}
+}
+
+type DirtyDumpChange struct {
+	TxHash  string           `json:"tx_hash"`
+	TxIndex uint             `json:"tx_index"`
```
Comment (markya0616) [inline core/state/dump.go]: Do we need tx index?

File: core/state/dump.go [hunk 2]
```diff
@@ -81,33 +81,81 @@ func (self *StateDB) Dump() []byte {
...
-			account.Storage[common.Bytes2Hex(key.Bytes())] = common.Bytes2Hex(value.Bytes())
+		if change.balanceChange > 0 {
+			balace := self.GetBalance(addr).String()
+			dumpChange.Change.Balance = &balace
 		}
-		dump.Accounts[common.Bytes2Hex(addr.Bytes())] = account
+		if change.storageChange > 0 {
+			stateObject := self.GetOrNewStateObject(addr)
+			for key, value := range stateObject.dirtyStorage {
+				dumpChange.Change.Storage[common.Bytes2Hex(key.Bytes())] = common.Bytes2Hex(value.Bytes())
```
Comment (markya0616) [inline core/state/dump.go]: Nil exception happens?

File: core/state/dump.go (hunk 1 above)
Comment (tailingchen) [reply core/state/dump.go]: removed in new version

File: core/state/dump.go (hunk 2 above)
Comment (tailingchen) [reply core/state/dump.go]: Yes, already fixed in new version.

**Review by markya0616 — APPROVED** [review]

**Pull request merged**